compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -2750,14 +2751,43 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
                             | LifetimeBinderKind::PolyTrait
                             | LifetimeBinderKind::WhereBound
                     );
-                    let (span, sugg) = if span.is_empty() {
+
+                    let (span, sugg, rm_poly_trait_span) = if span.is_empty() {
+                        let (generic_params, poly_trait_span, trait_ref_span) =
+                            if let Some(with_poly_trait_ref) =
+                                self.with_poly_trait_ref.get(&node_id)
+                                && higher_ranked
+                            {
+                                let generic_params = with_poly_trait_ref
+                                    .generic_param_idents
+                                    .iter()
+                                    .fold("".to_string(), |mut generic_params, x| {
+                                        generic_params += x.as_str();
+                                        generic_params += ", ";
+                                        generic_params
+                                    });
+                                (
+                                    generic_params,
+                                    with_poly_trait_ref.poly_trait,
+                                    with_poly_trait_ref.trait_ref,
+                                )
+                            } else {
+                                ("".to_string(), DUMMY_SP, DUMMY_SP)
+                            };
+
+                        let rm_poly_trait_span = if generic_params.is_empty() {
+                            DUMMY_SP
+                        } else {
+                            poly_trait_span.with_hi(trait_ref_span.lo())
+                        };
```

Comment:
if you use `rm_poly_trait_span` where you use `span` right now, you don't need to change any of the follow-up code, as you'd be replacing the old binder instead of generating a new one before the type.

Reply:
> if you use `rm_poly_trait_span` where you use `span` right now, you don't need to change any of the follow-up code, as you'd be replacing the old binder instead of generating a new one before the type.

Thank you. I need both rm_poly_trait_span and  span. But this code I wrote is indeed redundant and unclear. I simplified it.